lib/android/index.js

```diff
@@ -39,10 +44,30 @@ class Android {
     });
   }
 
+  async _downloadDir(sourcePath, destinationPath) {
+    const files = await this.client.readdir(this.id, sourcePath);
+
+    for (var file of files) {
+      const fullSourcePath = `${sourcePath}/${file.name}`; 
+      const fullDestinationPath = path.join(destinationPath, file.name);
+
+      if (file.isFile()) {
+        await this._downloadFile(fullSourcePath, fullDestinationPath);
```

Comment:
So, this is true, but it's subtle: you're using that `async` functions silently promote values to `Promise<T>`, and that nested promises are flattened.  By just `await`-ing, both branches (here and below) are simplified.

Reply:
re: nested promises being flattened, this only has an effect if you return Promise<Promise<T>> or return a Promise from an async function.

If you call an async function / function that returns a promise and do not await it, you are not guaranteed that it will finish by the time the function exits. You most definitely need to await.

Reply:
I guess I should have been more clear.  I didn't mean never wait for it, but the way it's written now you are serializing the download of each file.  If you collect the promises and move the await after the loop, then each download is asynchronous.